### ISSUE TITLE
Per-partition readers for consumer groups

### ISSUE THREAD

**Issue by abuchanan-nr** [segmentio/kafka-go#263]
(Apologies, I'm not sure exactly what to title this yet)

This relates to how best to scale message processors that are reading from a consumer group.

Ideally, each partition of a topic is processed completely in parallel, and on high-throughput topics this becomes critical. But how should we write and configure our services?

The simplest approach:
```go
func main() {
  reader := kafka.NewReader()
  for {
    msg := reader.FetchMessage()
    handle(msg)
    reader.CommitMessages(msg)
  }
}
```

If you have 12 partitions, then you start 12 instances of your service, et voilà.

But sometimes we have an in-memory cache that holds 20GB of data, while message processing only uses 0.5 CPU. It's wasteful to have 12 instances of a service. So we create a pool of readers:

```go
cache := NewBigCache()
for i := 0; i < N; i++ {
  go func() {
      reader := kafka.NewReader()
      for {
        msg := reader.FetchMessage()
        handle(cache, msg)
        reader.CommitMessages(msg)
      }
    }()
}
```

But, what is `N` here? How do you configure it? Is it `number of partitions / number of instances`? What if those numbers change? Should I write code to continuously check these numbers and scale our reader pool? These are questions I'm asking myself lately.

This also means it's more likely that the partitions won't be balanced evenly across the instances of the service (although maybe there's some clever fix for that involving a `GroupBalancer`).

We've also tried a pool of async worker threads:
```go
// create a pool of worker threads
work := make(chan Message)
for i := 0; i < 1000; i++ {
  startWorker(cache, work)
}

reader := NewReader()
for {
  msg := reader.FetchMessage()
  work <-msg
  // can't put this here anymore
  // reader.CommitMessages(msg)
}
```

...but that gets complex quickly, and messages can be committed out of order if you're not careful.

So what's the best solution? Maybe there's something missing from the kafka-go API to deal with this. Ideally, I think we want a single `Reader` per instance of a service, with a hook to be notified when a partition is assigned to that reader. 

I'm not sure what the API would look like, but here's what I imagine:
```go
group, err := NewConsumerGroup(conf)
group.Run(func(r *PartitionReader) {
  for {
    msg := r.FetchMessage()
    handle(cache, msg)
    r.CommitMessages(msg)
  }
})
```

`group.Run` would be called every time a new partition is assigned (i.e. every time a rebalance occurs).

An ancillary benefit is that each `PartitionReader` instance could be completely independent (separate connection and goroutine), so any potential bottleneck caused by [Reader.msgs](https://github.com/segmentio/kafka-go/blob/master/reader.go#L1193) could be avoided.

[Under the hood](https://github.com/segmentio/kafka-go/blob/master/reader.go#L1639) `Reader.start()` is already creating an internal connection + goroutine for each assigned partition, so it feels like all the structure is already there, we just need to find a nice way to expose this via the public API.

I've looked at shifting the `Reader` code around to allow for this, and I think it's doable. What do you guys think?

**Comment by stevevls** [comment]
I've been thinking about this as well, and it's something that I want to make happen.  I believe a more advanced user should be able to walk up to kafka-go and be able to choose between the "managed" consumer group as currently exported by `Reader` or to use a `ConsumerGroup` directly.  And if we get the API right, then we should also be able to easily refactor the `Reader` to use it. 😄 

Ideally, I think we should scope down the `ConsumerGroup` type to the mechanics of group membership and assignment and leave the actual consuming to higher level or client code.  While I like the simplicity of the last code snippet you proposed, I prefer the API client to manage their own go routines so that they can better orchestrate application lifecycle and/or do things that we didn't think of.  Something like:

```
type ConsumerGroupConfig struct {
    Topic string
     // all the other stuff...
}

type ConsumerGroup struct {}

func NewConsumerGroup(config ConsumerGroupConfig) (ConsumerGroup, error) {
    // kick off all the go routines.
}

func (cg *ConsumerGroup) NextAssignment(ctx context.Context) ([]int32, error) {
    // wait until next rebalance and return assignments.
}

func (cg *ConsumerGroup) Close() error {
    // tear down.
}
```

Then you would spawn a go routine to await changes in assignment and react accordingly.  WDYT?

FWIW, I don't think we should turn to the APIs that sarama exports for inspiration (either the CG impl in sarama itself or the sarama-cluster library) as I think they go too far in both hiding the underlying functionality and in imposing their processing model.

**Comment by stevevls** [comment]
Oh, and I totally forgot to mention offset management. 🤦‍♂  We'd want to account for that in the `ConsumerGroup` as well!

**Comment by abuchanan-nr** [comment]
> Ideally, I think we should scope down the ConsumerGroup type to the mechanics of group membership and assignment and leave the actual consuming to higher level or client code.

Sounds good.

It seems like you still need a separate consumer-group-specific partition reader type, in order to handle offset commits. In order to avoid clogging up the API with an advanced type, maybe a subpackage is warranted here? (root package for the quick and easy, subpackage for the more advanced users).

roughly:
```go
import (
  "github.com/segmentio/kafka-go/groups"
)

func main() {
  group, err := groups.NewConsumerGroup(conf)

  for {
    assignments, err := group.NextAssignments(ctx)
    for _, partition := range assignments {
      reader := groups.NewReader(partition, conf)
      go doWork(reader)        
    }
  }
}

func doWork(r *groups.Reader) {
  for {
    msg, err := r.FetchMessage()
    processMessage(msg)
    r.CommitMessages(msg)
  }
}
```

`groups.Reader` should have it's own connection, message queue, and commit queue.

**Comment by stevevls** [comment]
With all the rain this weekend, I ended up with a bunch of time to sit and write code.  I just put up a PR that adds the `ConsumerGroup` type, limited to just the membership and commit logic.  I put in sample code for how to use it to open a reader per-partition...I think that bit can live outside in order to keep the API surface small.  The PR is a WIP, but I think it captures the extent of what I hope to accomplish.  Would that solve your use case?

**Comment by stevevls** [comment]
Just merged #277, so this is do-able now!
